usr/opt/OSCX200/src/man/security.man
.\"
.\" *****************************************************************
.\" *                                                               *
.\" *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
.\" *                                                               *
.\" *   All Rights Reserved.  Unpublished rights  reserved  under   *
.\" *   the copyright laws of the United States.                    *
.\" *                                                               *
.\" *   The software contained on this media  is  proprietary  to   *
.\" *   and  embodies  the  confidential  technology  of  Digital   *
.\" *   Equipment Corporation.  Possession, use,  duplication  or   *
.\" *   dissemination of the software and media is authorized only  *
.\" *   pursuant to a valid written license from Digital Equipment  *
.\" *   Corporation.                                                *
.\" *                                                               *
.\" *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
.\" *   by the U.S. Government is subject to restrictions  as  set  *
.\" *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
.\" *   or  in  FAR 52.227-19, as applicable.                       *
.\" *                                                               *
.\" *****************************************************************
.\"
.\"
.\" HISTORY
.\"
.nr)S 12
.TH XSECURITY 1 "Release 5" "X Version 11"
.SH NAME
X Security \- X display access control
.SH SYNOPSIS
.PP
X provides mechanism for implementing many access control systems.
Release 5 includes four mechanisms:
.nf
.br
.ta 3.4i
    Host Access	Simple host-based access control.
    MIT-MAGIC-COOKIE-1	Shared plain-text "cookies".
    XDM-AUTHORIZATION-1	Secure DES based private-keys.
    SUN-DES-1	Based on Sun's secure rpc system.
.fi
.SH "ACCESS SYSTEM DESCRIPTIONS"
.IP "Host Access"
Any client on a host in the host access control list is allowed access to
the X server.  This system can work reasonably well in an environment
where everyone trusts everyone, or when only a single person can log in
to a given machine, and is easy to use when the list of hosts used is small.
This system does not work well when multiple people can log in to a single
machine and mutual trust does not exist.
The list of allowed hosts is stored in the X server and can be changed with
the \fIxhost\fP command.
When using the more secure mechanisms listed below, the host list is
normally configured to be the empty list, so that only authorized
programs can connect to the display.
.IP "MIT-MAGIC-COOKIE-1"
When using MIT-MAGIC-COOKIE-1,
the client sends a 128 bit "cookie"
along with the connection setup information.
If the cookie presented by the client matches one
that the X server has, the connection is allowed access.
The cookie is chosen so that it is hard to guess;
\fIxdm\fP generates such cookies automatically when this form of
access control is used.
The user's copy of
the cookie is usually stored in the \fI.Xauthority\fP file in the home
directory, although the environment variable \fBXAUTHORITY\fP can be used
to specify an alternate location.
\fIXdm\fP automatically passes a cookie to the server for each new
login session, and stores the cookie in the user file at login.
.IP
The cookie is transmitted on the network without encryption, so
there is nothing to prevent a network snooper from obtaining the data
and using it to gain access to the X server.  This system is useful in an
environment where many users are running applications on the same machine
and want to avoid interference from each other, with the caveat that this
control is only as good as the access control to the physical network.
In environments where network-level snooping is difficult, this system
can work reasonably well.
.IP "XDM-AUTHORIZATION-1"
For sites in the US, Release 5 contains a DES-based access control
mechanism called XDM-AUTHORIZATION-1.
It is similar in usage to MIT-MAGIC-COOKIE-1 in that a key is
stored in the \fI.Xauthority\fP file and is shared with the X server.
However,
this key consists of two parts - a 56 bit DES encryption key and 64 bits of
random data used as the authenticator.
.IP
When connecting to the X server, the application generates 192 bits of data
by combining the current time in seconds (since 00:00 1/1/1970 GMT) along
with 48 bits of "identifier".  For TCP/IP connections, the identifier is
the address plus port number; for local connections it is the process ID
and 32 bits to form a unique id (in case multiple connections to the same
server are made from a single process).  This 192 bit packet is then
encrypted using the DES key and sent to the X server, which is able to
verify if the requestor is authorized to connect by decrypting with the
same DES key and validating the authenticator and additional data.
This system is useful in many environments where host-based access control
is inappropriate and where network security cannot be ensured.
.IP "SUN-DES-1"
Recent versions of SunOS (and some other systems) have included a
secure public key remote procedure call system.  This system is based
on the notion of a network principal; a user name and NIS domain pair.
Using this system, the X server can securely discover the actual user
name of the requesting process.  It involves encrypting data with the
X servers public key, and so the identity of the user who started the
X server is needed for this; this identity is stored in the \fI.Xauthority\fP
file.  By extending the semantics of "host address" to include this notion of
network principal, this form of access control is very easy to use.
To allow access by a new user, use \fIxhost\fP.  For example,
.nf
    xhost keith@ joe@mit.edu
.fi
adds "keith" from the NIS domain of the local machine, and "joe" in
the "mit.edu" NIS domain.  For keith or joe to successfully connect
to the display, they must add the principal who started the server to
their \fI.Xauthority\fP file.  For example:
.nf
    xauth add expo.lcs.mit.edu:0 SUN-DES-1 unix.expo.lcs.mit.edu:x.lcs.mit.edu
.fi
This system only works on machines which support Secure RPC, and only for
users which have set up the appropriate public/private key pairs on their
system.  See the Secure RPC documentation for details.
.SH "THE AUTHORIZATION FILE"
.PP
Except for Host Access control, each of these systems uses data stored in
the \fI.Xauthority\fP file to generate the correct authorization information
to pass along to the X server at connection setup.  MIT-MAGIC-COOKIE-1 and
XDM-AUTHORIZATION-1 store secret data in the file; so anyone who can read
the file can gain access to the X server.  SUN-DES-1 stores only the
identity of the principal who started the server
(unix.\fIhostname\fP@\fIdomain\fP when the server is started by \fIxdm\fP),
and so it is not useful to anyone not authorized to connect to the server.
.PP
Each entry in the \fI.Xauthority\fP file matches a certain connection family
(TCP/IP, DECnet or local connections) and X display name (hostname plus display
number).  This allows multiple authorization entries for different displays
to share the same data file.  A special connection family (FamilyWild, value
65535) causes an entry to match every display, allowing the entry to be used
for all connections.  Each entry additionally contains the authorization
name and whatever private authorization data is needed by that authorization
type to generate the correct information at connection setup time.
.PP
The \fIxauth\fP program manipulates the \fI.Xauthority\fP file format.
It understands the semantics of the connection families and address formats,
displaying them in an easy to understand format.  It also understands that
SUN-DES-1 uses string values for the authorization data, and displays
them appropriately.
.PP
The X server (when running on a workstation) reads authorization
information from a file name passed on the command line with the \fI-auth\fP
option (see the \fIXserver\fP manual page).  The authorization entries in
the file are used to control access to the server.  In each of the
authorization schemes listed above, the data needed by the server to initialize
an authorization scheme is identical to the data needed by the client to
generate the appropriate authorization information, so the same file can be
used by both processes.  This is especially useful when \fIxinit\fP is used.
.IP "MIT-MAGIC-COOKIE-1"
This system uses 128 bits of data shared between the user and the X server.
Any collection of bits can be used.  \fIXdm\fP generates these keys using a
cryptographically secure pseudo random number generator, and so the key to
the next session cannot be computed from the current session key.
.IP "XDM-AUTHORIZATION-1"
This system uses two pieces of information.  First, 64 bits of random data,
second a 56 bit DES encryption key (again, random data) stored
in 8 bytes, the last byte of which is ignored.  \fIXdm\fP generates these keys
using the same random number generator as is used for MIT-MAGIC-COOKIE-1.
.IP "SUN-DES-1"
This system needs a string representation of the principal which identifies
the associated X server.  When \fIxdm\fP starts the X server, it uses the root
principal for the machine on which it is running
(unix.\fIhostname\fP@\fIdomain\fP, e.g.
"unix.expire.lcs.mit.edu@x.lcs.mit.edu").  Putting the correct principal
name in the \fI.Xauthority\fP file causes Xlib to generate the appropriate
authorization information using the secure RPC library.
.SH FILES
\&.Xauthority
.SH "SEE ALSO"
X(1), xdm(1), xauth(1), xhost(1), xinit(1), Xserver(1)


usr/opt/OSC200/src/usr/ccs/lib/libsecurity/security.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char *rcsid = "@(#)$RCSfile: security.c,v $ $Revision: 4.2.9.3 $ (DEC) $Date: 1993/04/08 19:21:59 $";
#endif
/*
 * @OSF_COPYRIGHT@
 */
/*
 * HISTORY
 * $OSF_Log:	security.c,v $
 * Revision 1.1.1.2  92/03/31  00:45:00  devrcs
 *  *** OSF1_1B25 version ***
 * 
 * Revision 1.7.2.2  1992/02/11  17:37:24  hosking
 * 	bug 4520: a bit more type widening (setluid() takes 'uid_t', not 'ushort')
 * 	[1992/02/11  17:36:29  hosking]
 *
 * Revision 1.5.2.3  91/03/11  15:42:33  seiden
 * 	Merge fixes up from 1.0.1
 * 	[91/03/11  15:26:06  seiden]
 * 
 * Revision 1.4.3.2  91/02/22  11:20:00  seiden
 * 	Build statpriv and chpriv under SEC_PRIV, not SEC_ARCH.
 * 
 * Revision 1.4  90/10/07  20:08:44  devrcs
 * 	Added EndLog Marker.
 * 	[90/09/28  20:17:20  gm]
 * 
 * Revision 1.3  90/07/17  12:21:20  devrcs
 * 	Internationalized
 * 	[90/07/05  07:44:32  staffan]
 * 
 * Revision 1.2  90/06/22  21:47:59  devrcs
 * 	Initial version from SecureWare
 * 	[90/05/31  11:41:55  staffan]
 * 
 * $OSF_EndLog$
 */
/*
 * Copyright (c) 1988, 1989 SecureWare, Inc.  All rights reserved.
 *
 * This Module contains Proprietary Information of SecureWare, Inc. and
 * should be treated as Confidential.
 */

/* #ident "@(#)security.c	6.4 08:38:17 8/21/91 SecureWare" */

/*LINTLIBRARY*/

#include <sys/secdefines.h>

/* #if SEC_BASE */ /*{*/

/*
 * This file contains a set of routines used to make programs
 * more secure.  Specifically, it contains system all interface
 * routines for security.
 */

#include <sys/types.h>

#include <sys/security.h>
#include <sys/audit.h>
#include <sys/sec_objects.h>
#if SEC_ARCH
#include <sys/secpolicy.h>
#endif
#include <prot.h>
#include <errno.h>

/*
 * Adjunct error number that allows analysis of security call failures.
 * The value of this external is set from the second byte of rval1.
 * The low order bits are used for the UNIX error number;
 * SEC_ERRNO_MASK is a mask of where in the errno field the sec_errno
 * value is placed.  SEC_ERRNO_SHIFT tells how far over to shift
 * the sec_errno value to get it right-justified.  Be careful if the
 * mask extends to the highest bit of errno and the machine sign extends
 * on a right shift.
 */

int sec_errno;
extern int errno;

#define SET_SEC_ERRNO()	{\
	sec_errno = (((long) errno) & SEC_ERRNO_MASK) >> SEC_ERRNO_SHIFT; \
	errno = ((long) errno) & ~SEC_ERRNO_MASK; \
	}

/*
 * The routines herein are user library stubs for the
 * security() sysent entry point.  The first argument to security() is
 * the system call code of the call and the remaining arguments
 * are call-specific.  Specify additional arguments as 0.
 *
 * Note that some return actual values (getluid(), getpriv())
 * while the others return 0 for success.  All return -1 for failure.
 */

/*
 * Some of the syscall-handling can be short-circuited if we know that
 * we have a kernel without SEC_BASE.
 */
static int sec_conf_known = 0, sec_conf_stat;

/*
 * Force all open files in the system to the filename provided to become
 * invalid.  On the next read/write/ioctl access, send a signal to the
 * effected process and return an error.
 */
int
stopio(filename)
	caddr_t filename;
{
	register int ret;

	if (security_is_on())
		ret = security(SEC_STOPIO, filename, NULL, NULL, NULL, NULL);
	else
		ret = revoke(filename);
	if (ret != -1)
		ret = 0;
	else
		SET_SEC_ERRNO();

	return ret;
}


/*
 * Get the system's notion of the login UID associated with the process.
 * If it has not been set yet, return an error.
 */
int
getluid()
{
	register int ret;

	ret = security(SEC_GETLUID, NULL, NULL, NULL, NULL, NULL);

	if (ret == -1)
		SET_SEC_ERRNO();

	return ret;
}


/*
 * Set the system's notion of the login UID for this process and all its
 * descendants.  Once set, it cannot be reset.
 */
int
setluid(uid)
	uid_t uid;
{
	register int ret;

	ret = security(SEC_SETLUID, uid, NULL, NULL, NULL, NULL);
	if (ret != -1)
		ret = 0;
	else
		SET_SEC_ERRNO();

	return ret;
}


/*
 * Get the system privileges for this process and store them in the supplied
 * buffer.  The size in bytes of the number of privileges is returned.
 * Privtype specifies either the maximum, base or effective privileges.
 */
int
getpriv(privtype, vec)
	long privtype;
	register priv_t *vec;
{
	obj_t objid;
	register int i;
	register uid_t ruid, euid;
	static privvec_t allprivs;
	static int done_once = 0;

	objid.o_pid = 0;

	if (security_is_avail())
		return statprivsys(privtype, vec, OT_PROCESS, &objid);
	(void) memset(vec, 0, sizeof(*vec));
	ruid = getuid();
	euid = geteuid();
	if (!ruid || !euid) {
		if (!done_once) {
			for (i = 0;  i <= SEC_MAX_SPRIV;  i++)
				ADDBIT(vec, i);
			(void) memmove(allprivs, vec, sizeof allprivs);
			done_once = 1;
		}
		else
			(void) memmove(vec, allprivs, sizeof allprivs);
	}
	return 0;
}


/*
 * Reset the system privileges for this process.  They cannot be more
 * than was already set.  Privtype specifies either the maximum,
 * base or effective privileges.
 */
int
setpriv(privtype, vec)
	long privtype;
	register priv_t *vec;
{
	obj_t objid;
	register int i;
	register int ruid, euid;

	objid.o_pid = 0;

	if (security_is_avail())
		return chprivsys(privtype, vec, OT_PROCESS, &objid);
	ruid = getuid();
	euid = geteuid();
	if (!ruid || !euid)
		return 0;
	for (i = 0;  i < SEC_SPRIVVEC_SIZE;  i++) {
		if (vec[i]) {
			errno = EPERM;
			SET_SEC_ERRNO();
			return -1;
		}
	}
	return 0;
}

/*
 * check if the system has security code turned on or not
 */
int
security_is_on()
{
	register int ret;

	if (!sec_conf_known)
		(void) security_is_avail();
	if (sec_conf_known && !sec_conf_stat)
		return 0;

	ret = security(SEC_SWITCH_CALL, SEC_SWITCH_STAT, NULL, NULL, NULL, NULL);

	if (ret == -1)
		SET_SEC_ERRNO();

	return ret;
}

/*
 * check if the system was built with SEC_BASE turned on
 */
int
security_is_avail()
{
	register int ret;

	if (sec_conf_known)
		return sec_conf_stat;
	ret = security(SEC_SWITCH_CALL, SEC_SWITCH_CONF, NULL, NULL, NULL, NULL);

	if (ret == -1) {
		SET_SEC_ERRNO();
	}
	else {
		sec_conf_known = 1;
		sec_conf_stat = ret;
	}

	return ret;
}

/*
 * turn security code on
 */
int
security_turn_on(gid)
	gid_t gid;
{
	register int ret;

	ret = security( SEC_SWITCH_CALL, SEC_SWITCH_ON, gid, NULL, NULL, NULL);

	if (ret == -1)
		SET_SEC_ERRNO();

	return ret;
}
/*
 * turn security code off
 */
int
security_turn_off()
{
	register int ret;

	ret = security( SEC_SWITCH_CALL, SEC_SWITCH_OFF, NULL, NULL, NULL, NULL);

	if (ret == -1)
		SET_SEC_ERRNO();

	return ret;
}

#if SEC_PRIV /*{*/
/*
 * Retrieve the specified privilege set for a file.
 * Privtype is one of SEC_POTENTIAL_PRIV or SEC_GRANTED_PRIV
 */

int
statpriv(file, privtype, vec)
	caddr_t file;
	long privtype;
	priv_t *vec;
{
	obj_t objid;

	objid.o_file = file;

	return statprivsys(privtype, vec, OT_REGULAR, &objid);
}

/*
 * Set the privilege set for a file.
 * Privtype is one of SEC_POTENTIAL_PRIV or SEC_GRANTED_PRIV
 */

chpriv(file, privtype, vec)
	caddr_t file;
	long privtype;
	priv_t *vec;
{
	obj_t objid;

	objid.o_file = file;

	return chprivsys(privtype, vec, OT_REGULAR, &objid);
}
#endif /*} SEC_PRIV */

/*
 * Obtain the privileges associated with an object, whether the object be
 * a process, file or IPC entity.
 */
int
statprivsys(privtype, vec, objtype, objid)
	long privtype;
	priv_t *vec;
	long objtype;
	obj_t *objid;
{
	register int ret;

	ret = security( SEC_STATPRIV, privtype, vec, objtype, objid, NULL);

	if (ret == -1)
		SET_SEC_ERRNO();

	return ret;
}


/*
 * Change the privileges associated with an object, whether the object be
 * a process, file or IPC entity.
 */
int
chprivsys(privtype, vec, objtype, objid)
	long privtype;
	priv_t *vec;
	long objtype;
	obj_t *objid;
{
	register int ret;

	ret = security( SEC_CHPRIVSYS, privtype, vec, objtype, objid, NULL);
	if (ret != -1)
		ret = 0;
	else
		SET_SEC_ERRNO();

	return ret;
}


#if SEC_ARCH
/*
 * Obtain the effective access for the object, considering ALL security
 * policies and UNIX constraints.
 */
int
eaccess(file, mode)
	caddr_t file;
	long mode;
{
	register int ret;

	ret = security(SEC_EACCESS, file, mode, NULL, NULL, NULL);
	if (ret != -1)
		ret = 0;
	else
		SET_SEC_ERRNO();

	return ret;
}
#endif

#if SEC_MAC /*{*/
/*
 * Make a multilevel directory.
 */
int
mkmultdir(file, mode)
	caddr_t file;
	int mode;
{
	register int ret;

	ret = security(SEC_MKMULTDIR, file, NULL, NULL, NULL, NULL);
	if (ret != -1)
		ret = 0;
	else
		SET_SEC_ERRNO();

	return ret;
}


/*
 * Remove a multilevel directory.
 */
int
rmmultdir(file, mode)
	caddr_t file;
	int mode;
{
	register int ret;

	ret = security(SEC_RMMULTDIR, file, NULL, NULL, NULL, NULL);
	if (ret != -1)
		ret = 0;
	else
		SET_SEC_ERRNO();

	return ret;
}


/*
 * Query a file to see if it is a multilevel directory.  It returns 1
 * if it is a multilevel directory, 0 if not, and -1 for errors.
 */
int
ismultdir(file, mode)
	caddr_t file;
	int mode;
{
	register int ret;

	ret = security(SEC_ISMULTDIR, file, NULL, NULL, NULL, NULL);

	if (ret == -1)
		SET_SEC_ERRNO();

	return ret;
}
#endif /*} SEC_MAC */



#if SEC_ARCH /*{*/

/*
 * Get a security label from one of the policies.
 */
int
getlabel(policy,tagnum,attr,objtype,objid)
	ulong_t	policy, tagnum;
	attr_t *attr;
	long objtype;
	obj_t *objid;
{
	register int ret;

	ret = security(SEC_GETLABEL, policy, tagnum, attr, objtype, objid);
	if (ret != -1)
		ret = 0;
	else
		SET_SEC_ERRNO();

	return ret;
}


/*
 * Set a security label from one of the policies.
 */
int
setlabel(policy, tagnum, attr, objtype, objid)
	ulong_t policy, tagnum;
	attr_t *attr;
	long objtype;
	obj_t *objid;
{
	register int ret;

	ret = security(SEC_SETLABEL, policy, tagnum, attr, objtype, objid);
	if (ret != -1)
		ret = 0;
	else
		SET_SEC_ERRNO();

	return ret;
}


/*
 * Label mount an insecure file system type, assigning all files therein to
 * single values (fixed by the labels) for the security policies.
 */
int
lmount(mnttype, name, flags, argp, attrs)
	caddr_t	mnttype;
	caddr_t	name;
	long flags;
	caddr_t argp;
	attr_t *attrs;
{
	register int ret;

	ret = security(SEC_LMOUNT, mnttype, name, flags, argp, attrs);
	if (ret != -1)
		ret = 0;
	else
		SET_SEC_ERRNO();

	return ret;
}

/*
 * Return TRUE if the file system supports labels and/or privileges
 */
int
islabeledfs(file)
	caddr_t file;
{
	register int ret;

	ret = security(SEC_ISLABELEDFS, file, NULL, NULL, NULL, NULL);

	if (ret == -1)
		SET_SEC_ERRNO();

	return ret;
}
#endif /*} SEC_ARCH */
/* #endif */ /*} SEC_BASE */
